eos_data_distribution/avahi_monitor.py

```diff
@@ -79,23 +96,102 @@ def __init__(self):
-        ifname = sda.siocgifname(interface)
-        print "Found Service data for service '%s' of type '%s' (%s) in domain '%s' on %s.%i:" % (name, h_type, type, domain, ifname, protocol)
+        logger.debug(
+            "Found Service data for service '%s' of type '%s' (%s) in domain '%s' on %s.%i:",
+                     name, h_type, type, domain, ifname, protocol)
+
+        self.refresh_network()
+
+        if address in self.ips:
+            return
+
         faceUri = face_uri_from_triplet(type, host, port)
-        check_call(["nfdc", "add-nexthop",
-                    "-c", str(defaults.RouteCost.LOCAL_NETWORK),
-                    str(SUBSCRIPTIONS_SOMA), faceUri])
-        self._registry[build_registry_key(name, type, domain)] = faceUri
+        key = build_registry_key(name, type, domain)
+        self._registry[key] = faceUri
+
+        if address in self.gateways:
+            self._ndn_gateways[key] = address
+
+        prev_routed = self._routed
+        self.update_routed_status()
+
+        if not self._routed and not prev_routed:
+            # not routed, wasn't routed, just add this new link
+            self.add_nexthop(faceUri)
+        else:
+            # all the rest handled by generic code
+            self.process_route_changes(prev_routed)
 
     def service_removed_cb(self, sda, interface, protocol, name, type, domain, flags):
-        ifname = sda.siocgifname(interface)
-        print "Disappeared Service '%s' of type '%s' in domain '%s' on %s.%i." % (name, type, domain, ifname, protocol)
-        faceUri = self._registry[build_registry_key(name, type, domain)]
-        check_call(
-            ["nfdc", "remove-nexthop", str(SUBSCRIPTIONS_SOMA), faceUri])
+        logger.debug(
+            "Disappeared Service '%s' of type '%s' in domain '%s' on %s.%i.",
+            name, type, domain, ifname, protocol)
+
+        self.refresh_network()
+
+        key = build_registry_key(name, type, domain)
+        try:
+            del self._ndn_gateways[key]
+        except KeyError:
+            pass
+
+        faceUri = self._registry[key]
+        del self._registry[key]
+
+        prev_routed = self._routed
+        self.update_routed_status()
+
+        if not self._routed and not prev_routed:
+            # not routed, wasn't routed, just remove this link
+            self.remove_nexthop(faceUri)
+        else:
+            # all the rest handled by generic code
+            self.process_route_changes()
+
+    def add_nexthop(self, faceUri):
+        return check_call(["nfdc", "add-nexthop",
+                           "-c", str(defaults.RouteCost.LOCAL_NETWORK),
+                           str(SUBSCRIPTIONS_SOMA), faceUri])
+
+    def remove_nexthop(self, faceUri):
+        return check_call(["nfdc", "remove-nexthop",
+                           str(SUBSCRIPTIONS_SOMA), faceUri])
+
+    def refresh_network(self, *a, **kw):
```

Comment:
it will emit them when it goes down on an interface it was monitoring, that's what we're interested in. and when an interface comes up, if avahi handles it, it will start emiting `service-added`

so we're covered: we're only interested in interfaces avahi is monitoring anyway